src/type_hash.rs

```diff
@@ -0,0 +1,220 @@
+use crate::prelude::*;
+use lazy_static::lazy_static;
+use std::any::TypeId;
+use std::collections::{BTreeMap, HashMap};
+use std::sync::RwLock;
+
+// (SPEC) The type of a struct is encoded as name ‖ "(" ‖ member₁ ‖ "," ‖
+// member₂ ‖ "," ‖ … ‖ memberₙ ")" where each member is written as type ‖ " " ‖
+// name. For example, the above Mail struct is encoded as Mail(address
+// from,address to,string contents)
+pub fn encode_type<T: StructType>(value: &T) -> String {
+    let mut builder = TypeHashBuilder {
+        outer: None,
+        inner: BTreeMap::new(),
+    };
+
+    value.add_members(&mut builder);
+
+    let mut buffer = String::new();
+    let outer = builder.outer.unwrap();
+    assert!(outer.name == T::TYPE_NAME);
+
+    fn add_type(s: &mut String, t: &EncodedType) {
+        s.push_str(t.name);
+        s.push_str("(");
+        let mut members = t.members.iter();
+        if let Some(member) = members.next() {
+            member.write(s);
+        }
+        for member in members {
+            s.push_str(",");
+            member.write(s);
+        }
+        s.push_str(")");
+    }
+
+    add_type(&mut buffer, &outer);
+    for inner in builder.inner.values() {
+        add_type(&mut buffer, inner);
+    }
+    buffer
+}
+
+lazy_static! {
+    static ref CACHE: RwLock<HashMap<TypeId, Bytes32>> = RwLock::new(HashMap::new());
+}
+
+/// Memoized type hash
+pub fn type_hash<T: StructType>(value: &T) -> Bytes32 {
+    let read = CACHE.read().unwrap();
+    if let Some(cached) = read.get(&TypeId::of::<T>()) {
+        return *cached;
+    }
+    drop(read);
+
+    // (SPEC) keccak256(encodeType(typeOf(s)))
+    let encoded = encode_type(value);
+    let result = keccak(encoded.as_bytes());
+
+    let mut write = CACHE.write().unwrap();
+    write.insert(TypeId::of::<T>(), result);
+    result
+}
+
+pub struct TypeHashBuilder {
+    // (SPEC) If the struct type references other struct types (and these in
+    // turn reference even more struct types), then the set of referenced struct
+    // types is collected, sorted by name and appended to the encoding. An
+    // example encoding is Transaction(Person from,Person to,Asset
+    // tx)Asset(address token,uint256 amount)Person(address wallet,string name).
+    //
+    // NOTE: This means that the outer type gets special treatment, since it is not part
+    // of the sorted set.
+    outer: Option<EncodedType>,
+    inner: BTreeMap<&'static str, EncodedType>,
+}
+
+impl TypeHashBuilder {
+    fn get_encoded_type_mut(&mut self, name: &'static str) -> Option<&mut EncodedType> {
+        if let Some(outer) = &self.outer {
```

Comment:
You can go all the way with...

```
let inner = &mut self.inner;
self.outer
    .as_mut()
    .filter(|o| o.name == name)
    .or_else(move || inner.get_mut(name))
```

But is this style really better? It may be old-school, but I just prefer the simple if-statement.

Reply:
That's fine, I guess I have a pet peeve with how `if let` and boolean conditions not always composing nicely. Lengthy Rust RFC discussions have been had on generalizing `if let`, but nothing happened :disappointed:.

Reply:
My preference for using the simple if/let is probably more historical. It used to be that I used C# extensively, which has something called `linq` which enables a functional style in coding. But, in C#, JS, and other languages these constructs (and really anything involving closures) come with significant performance penalties related to heap allocations and black boxing the optimizer. So, I've had to go and clear out every instance of the use of these constructs and just lower them to for loops and if statements. I learned that at least half the time, the actual code was shorter and easier to understand after the transformation. After a while, it became tedious to ask "Am I in performance-critical code right now? Yes -> use a loop, No -> use linq" and instead just use for loops exclusively for consistency.

I'm trying to be more open-minded about it now that with Rust performance is not nearly as much of a concern.